Extract TestHelpers into own test target

DavidFair opened the pull request:
**Description of work.**
Extracts TestHelpers into it's own test target for a number of reasons:

- Avoid redundant compilation
- Automated tools (clang-tidy) will parse these files once instead of n
times
- Helps MSVC find the various targets
- Makes dep chain explicit.

This also includes some minor modernisation to the CMakelists
associated, such as replacing include_directories with
target_include_directories.

Long term we need to do the following:
- Split out Qt and Framework only helpers
- Consolidate as much implementation into the associated *.cpp files
- Move target specific code (e.g. Muon) into their associated targets
<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**
- Check CI builds, this will cause compilation errors if it's incorrect

There is no associated issue. - Part of the clang tidy work

*This does not require release notes* because it's an internal change that has no visible side-effects (apart from faster compilation)

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Extract TestHelpers into own test target

Extracts TestHelpers into it's own test target for a number of reasons:

- Avoid redundant compilation
- Automated tools (clang-tidy) will parse these files once instead of n
times
- Helps MSVC find the various targets
- Makes dep chain explicit.

This also includes some minor modernisation to the CMakelists
associated, such as replacing include_directories with
target_include_directories.

Long term we need to do the following:
- Split out Qt and Framework only helpers
- Consolidate as much implementation into the associated *.cpp files
- Move target specific code (e.g. Muon) into their associated targets

DavidFair committed:
Split up Framework test helpers and Qt test helpers

This goes hand-in-hand with the Qt / Framework seperation going on as
part of the Conda work

DavidFair committed:
Handle dynamic factory registration

Use an object library instead of a static lib. This way we get the
benefits of a static lib (built once use many), whilst still linking in
the object files to the main executable.

This requires a bit of clunky usage so we wrap this all in a function

DavidFair committed:
Fix test which relied on broken linkage

This test was marked as expecting to fail, since we didn't handle the
static initalization problem properly.

By fixing it the factory can now find the type causing the test to
mis-fail

DavidFair committed:
Add missing gmock depedency to TestHelpers

Adds missing gmock which was only noticed using a pending PR

Harrietbrown reviewed (dismissed):
All builds are working without issue. Good job.

DavidFair committed:
Fix new usages of test helpers that changed upstream

This fixes various usages of test helpers that did not exist upstream when this PR was started

DavidFair committed:
Fix OSX deps which included duplicate symbols

DavidFair committed:
Fix warnings now detected by cppcheck

Fixes various warnings that were previously missed, as the headers / cpp
files weren't explicitly known to CMake

DavidFair committed:
Run CMake Format over changes

CMake Format was enabled mid-branch, so run a pass over the files I've
changed

DavidFair committed:
Fixup new usages since last rebase

Changes MantidTestHelpers to MantidFrameworkTestHelpers in the new
places this is used

Harrietbrown reviewed (approved):
All builds are working without issue. Good job.

gemmaguest commented on Framework/API/test/RemoteJobManagerFactoryTest.h:
```diff
@@ -149,16 +149,7 @@ class RemoteJobManagerFactoryTest : public CxxTest::TestSuite {
-                          Mantid::API::RemoteJobManagerFactory::Instance().create("Fermi"),
-                      const Mantid::Kernel::Exception::NotFoundError &);
-
+    TS_ASSERT(Mantid::API::RemoteJobManagerFactory::Instance().create("Fermi"));
```
I'm not clear why this test has changed?

DavidFair replied:
Previously this test failed because it (incorrectly) relied on the Mantid Framework not loading in correctly
This is because the linker will strip out unused shared objects if we don't explicitly statically init them. I've got a workaround in the PR that fixes this for all cases, which means that alg actually runs
